go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go

```diff
@@ -375,34 +418,124 @@ func (c *Controller) repair() error {
 		}
 
 		node := c.nodes[svc.node]
-		if item.Nexthops[0] != node.v4MgmtIP.String() && item.Nexthops[0] != node.v6MgmtIP.String() {
+		nextHopV4 := node.v4MgmtIP.String()
+		nextHopV6 := node.v6MgmtIP.String()
+		if config.OVNKubernetesFeature.EnableInterconnect {
+			zoneKnown, svcNodeInLocalZone := c.nodeZoneState(node.name)
+			if !zoneKnown {
+				klog.Errorf("Failed to verify whether the svc node: %s is in the local zone, deleting lrp", node.name)
+				return true
+			}
+			if !svcNodeInLocalZone {
+				nextHopV4 = node.transitIPV4.String()
+				nextHopV6 = node.transitIPV6.String()
+			}
+		}
+
+		if item.Nexthops[0] != nextHopV4 && item.Nexthops[0] != nextHopV6 {
 			klog.Infof("Egress service repair will delete %s because it is uses a stale nexthop for service %s: %v", logicalIP, svcKey, item)
 			return true
 		}
 
 		if utilnet.IsIPv4String(logicalIP) {
-			svcKeyToConfiguredV4Endpoints[svcKey] = append(svcKeyToConfiguredV4Endpoints[svcKey], logicalIP)
+			svcKeyToLocalConfiguredV4Endpoints[svcKey] = append(svcKeyToLocalConfiguredV4Endpoints[svcKey], logicalIP)
 			return false
 		}
 
-		svcKeyToConfiguredV6Endpoints[svcKey] = append(svcKeyToConfiguredV6Endpoints[svcKey], logicalIP)
+		svcKeyToLocalConfiguredV6Endpoints[svcKey] = append(svcKeyToLocalConfiguredV6Endpoints[svcKey], logicalIP)
 		return false
 	}
 
 	errorList := []error{}
-	err = libovsdbops.DeleteLogicalRouterPoliciesWithPredicate(c.nbClient, ovntypes.OVNClusterRouter, p)
+	ops := []libovsdb.Operation{}
+	ops, err = libovsdbops.DeleteLogicalRouterPolicyWithPredicateOps(c.nbClient, ops, ovntypes.OVNClusterRouter, lrpPredicate)
 	if err != nil {
 		errorList = append(errorList,
 			fmt.Errorf("error deleting stale logical router policies from router %s: %v", ovntypes.OVNClusterRouter, err))
 	}
 
+	if config.OVNKubernetesFeature.EnableInterconnect {
+		lrsrPredicate := func(item *nbdb.LogicalRouterStaticRoute) bool {
+			svcKey, found := item.ExternalIDs[svcExternalIDKey]
+			if !found {
+				return false
+			}
+
+			if item.Policy == nil || *item.Policy != nbdb.LogicalRouterStaticRoutePolicySrcIP {
+				klog.Infof("Egress service repair will delete lrsr for service %s it has an invalid policy: %v", svcKey, item)
+				return true
+			}
+
+			svc, found := c.services[svcKey]
+			if !found {
+				klog.Infof("Egress service repair will delete lrsr for service %s because it is no longer a valid egress service: %v", svcKey, item)
+				return true
+			}
+
+			node := c.nodes[svc.node]
+			zoneKnown, svcNodeInLocalZone := c.nodeZoneState(node.name)
```

Comment:
This is confusing because a map would return assignments in the opposite order, but you created a function that returns them like this:
// nodeZoneState returns whether the node zone is known and if it is local to the zone controller
// should only be called with controller locked
func (c *Controller) nodeZoneState(nodeName string) (exists bool, isNodeZoneLocal bool) {
	isNodeZoneLocal, exists = c.nodesZoneState[nodeName]
	return
}

why not just directly access the map instead of using a function?

Reply:
good point, there is no point to have that function.